src/main/scala/com/github/pathikrit/dijon/package.scala

```diff
@@ -135,71 +131,93 @@ package object dijon {
     override def hashCode: Int = underlying.hashCode
   }
 
-  def compact(json: SomeJson): String = writeToString[SomeJson](json)
+  def compact(json: SomeJson): String = new String(writeToArray[SomeJson](json), UTF_8)
 
-  def pretty(json: SomeJson): String = writeToString[SomeJson](json, WriterConfig.withIndentionStep(2))
+  def pretty(json: SomeJson): String = new String(writeToArray[SomeJson](json, prettyConfig), UTF_8)
 
-  def parse(s: String): SomeJson = readFromString[SomeJson](s)
+  def parse(s: String): SomeJson = readFromArray[SomeJson](s.getBytes(UTF_8))
 
   implicit class JsonStringContext(val sc: StringContext) extends AnyVal {
     def json(args: Any*): SomeJson = parse(sc.s(args: _*))
   }
 
   implicit val codec: JsonValueCodec[SomeJson] = new JsonValueCodec[SomeJson] {
-    override def decodeValue(in: JsonReader, default: SomeJson): SomeJson = (in.nextToken(): @switch) match {
-      case 'n' => in.readNullOrError(default, "expected `null` value")
-      case '"' =>
+    override def decodeValue(in: JsonReader, default: SomeJson): SomeJson = decode(in, maxParsingDepth)
+
+    override def encodeValue(x: SomeJson, out: JsonWriter): Unit = encode(x, out, maxSerializationDepth)
+
+    override val nullValue: SomeJson = None
+
+    private[this] def decode(in: JsonReader, depth: Int): SomeJson = {
+      val b = in.nextToken()
+      if (b == 'n') in.readNullOrError(None, "expected `null` value")
+      else if (b == '"') {
         in.rollbackToken()
         in.readString(null)
-      case 'f' | 't' =>
+      } else if (b == 't' || b == 'f') {
         in.rollbackToken()
         in.readBoolean()
-      case '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '-' =>
+      } else if ((b >= '0' && b <= '9') || b == '-') {
         in.rollbackToken()
         val d = in.readDouble()
         val i = d.toInt
         if (i.toDouble == d) i
         else d
-      case '[' =>
-        val arr = new mutable.ArrayBuffer[SomeJson]
+      } else if (b == '[') {
+        if (depth <= 0) in.decodeError("depth limit exceeded")
+        val arr = new mutable.ArrayBuffer[SomeJson](initArrayCapacity)
         if (!in.isNextToken(']')) {
           in.rollbackToken()
-          do arr += decodeValue(in, default)
+          val dp = depth - 1
+          do arr += decode(in, dp)
           while (in.isNextToken(','))
           if (!in.isCurrentToken(']')) in.arrayEndOrCommaError()
         }
         arr
-      case '{' =>
-        val obj = new util.LinkedHashMap[String, SomeJson]
+      } else if (b == '{') {
+        if (depth <= 0) in.decodeError("depth limit exceeded")
+        val obj = new util.LinkedHashMap[String, SomeJson](initMapCapacity)
```

Comment:
The smaller initial capacity could backfire for objects with larger numbers of keys due to more resizing/hashing. For example, I'd expect this to perform worse for objects with 16 keys than it did previously.

I've been wondering whether it might be advantageous to buffer these into an array prior to constructing the `LinkedHashMap` to avoid the rehashing.

A side benefit of knowing the exact size before constructing the `LinkedHashMap` is that you could make the final object compact, e.g. you could size the capacity to exactly 5. For big objects, this would heap with peak live set size. It might help with small objects, too, but I'm less sure about that.

I haven't actually benchmarked anything, so no idea whether it's worth it or not. Just a thought.